# 과제 6/30

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

In [13]:
import pandas as pd
import seaborn as sns  # 데이타셋 가져올라고

df = sns.load_dataset('titanic')
df.head()  
# pclass 나 class 중 제거해야됨 하나는 똑같은 데이터임


survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

### Q1. titanic 데이터셋에서 age, fare 2개 열을 선택하여 데이터프레임 df를 만드세요

In [17]:

df1 = pd.concat([df['age'],df['fare']], axis=1)
df1

age     fare
0    22.0   7.2500
1    38.0  71.2833
2    26.0   7.9250
3    35.0  53.1000
4    35.0   8.0500
..    ...      ...
886  27.0  13.0000
887  19.0  30.0000
888   NaN  23.4500
889  26.0  30.0000
890  32.0   7.7500

[891 rows x 2 columns]

### Q2. 10을 더하는 함수 add_10 과 두 객체의 합 add_two_obj 함수를 정의한 후 결과값이 각각 20과 20을 출력하세요.

In [7]:
def add_10(a):
    return a+10

def add_two_obj(a,b):
    return a+b

print(add_10(10))
print(add_two_obj(10,10))

20
20


### Q3. 10을 곱하는 사용자 함수와 두 객체의 나눗셈을 수행하는 사용자 함수를 정의한 후 결과값이 각각 200과 10을 출력하세요.

In [8]:
def mul_10(a):
    return a*10

def div(a,b):
    return a//b

print(mul_10(20))
print(div(20,2))
    

200
10


### Q4. df.age 에 add_10 함수를 적용해서 출력하세요.

In [16]:
add_10(df[['age']])

age
0    32.0
1    48.0
2    36.0
3    45.0
4    45.0
..    ...
886  37.0
887  29.0
888   NaN
889  36.0
890  42.0

[891 rows x 1 columns]

### Q5. 시리즈 객체 + 숫자를 연산하는 방식으로 df.age에 10을 더한 후 sr2에 저장하고 출력하세요.

In [12]:
df['sr2'] = df.age + 10
sr2

0      32.0
1      48.0
2      36.0
3      45.0
4      45.0
       ... 
886    37.0
887    29.0
888     NaN
889    36.0
890    42.0
Name: age, Length: 891, dtype: float64

### Q6. 람다 함수를 활용해서 위와 같은 연산을 수행하고 결과를 sr3로 출력하세요.

In [17]:
sr3 = lambda x:x+10 , df.age
sr3

(<function __main__.<lambda>(x)>,
 0      22.0
 1      38.0
 2      26.0
 3      35.0
 4      35.0
        ... 
 886    27.0
 887    19.0
 888     NaN
 889    26.0
 890    32.0
 Name: age, Length: 891, dtype: float64)

### Q7. df의 'age', 'fare' 칼럼에 10을 덧셈연산하는 사용자 함수를 정의하고 applymap()으로 각 사용자 함수를 매핑한 후 출력하세요.

In [7]:
def add_10(n):
    return n+10


#df_map = df[['age','fare']].applymap(add)
df_map = df1.applymap(add_10)
print(df_map)


      age     fare
0    32.0  17.2500
1    48.0  81.2833
2    36.0  17.9250
3    45.0  63.1000
4    45.0  18.0500
..    ...      ...
886  37.0  23.0000
887  29.0  40.0000
888   NaN  33.4500
889  36.0  40.0000
890  42.0  17.7500

[891 rows x 2 columns]


### Q8. df_map의 [0,0]을 NaN으로 변경한 후 missing_value 함수를 적용해서 불린 시리즈를 반환하세요.

In [8]:
import numpy as np
df_map.iloc[0,0] = np.nan # null값 대입
def missing_value(series):
    return series.isnull()
result = df_map.apply(missing_value, axis=0)

print(result)

#뭐야이게 애바네

       age   fare
0     True  False
1    False  False
2    False  False
3    False  False
4    False  False
..     ...    ...
886  False  False
887  False  False
888   True  False
889  False  False
890  False  False

[891 rows x 2 columns]


### Q9. titanic 데이터셋을 seaborn에서 로딩한 후 블린 인덱싱을 이용하여 아래 사항들을 수행하세요.
- 나이가 10대(10~19세)인 승객만 따로 선택
- 나이가 10세 미만(0~9세)이고 여성인 승객만 따로 선택
- 나이가 10세 미만(0~9세) 또는 60세 이상인 승객의 age, sex, alone 열만 선택

In [8]:
import seaborn as sns
titanic = sns.load_dataset('titanic')
titanic.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [11]:
mask1 = (titanic['age']>=10)&(titanic['age']<20)
titanic1= titanic.loc[mask1,:]
titanic1.head()

survived  pclass     sex   age  sibsp  parch      fare embarked   class  \
9          1       2  female  14.0      1      0   30.0708        C  Second   
14         0       3  female  14.0      0      0    7.8542        S   Third   
22         1       3  female  15.0      0      0    8.0292        Q   Third   
27         0       1    male  19.0      3      2  263.0000        S   First   
38         0       3  female  18.0      2      0   18.0000        S   Third   

      who  adult_male deck  embark_town alive  alone  
9   child       False  NaN    Cherbourg   yes  False  
14  child       False  NaN  Southampton    no   True  
22  child       False  NaN   Queenstown   yes   True  
27    man        True    C  Southampton    no  False  
38  woman       False  NaN  Southampton    no  False

In [12]:
mask2 = (titanic['age']<10)&(titanic['sex']=='female')
titanic2 = titanic.loc[mask2,:]
titanic2.head()

survived  pclass     sex  age  sibsp  parch     fare embarked   class  \
10          1       3  female  4.0      1      1  16.7000        S   Third   
24          0       3  female  8.0      3      1  21.0750        S   Third   
43          1       2  female  3.0      1      2  41.5792        C  Second   
58          1       2  female  5.0      1      2  27.7500        S  Second   
119         0       3  female  2.0      4      2  31.2750        S   Third   

       who  adult_male deck  embark_town alive  alone  
10   child       False    G  Southampton   yes  False  
24   child       False  NaN  Southampton    no  False  
43   child       False  NaN    Cherbourg   yes  False  
58   child       False  NaN  Southampton   yes  False  
119  child       False  NaN  Southampton    no  False

In [21]:
mask3 = (titanic['age']<10)|(titanic['age']>=60), ['age','sex','alone']
titanic3 = titanic.loc[mask3]
titanic3.head()

age     sex  alone
7    2.0    male  False
10   4.0  female  False
16   2.0    male  False
24   8.0  female  False
33  66.0    male   True

### Q10. 나이가 10세 미만(0~9세) 또는 60세 이상인 승객의 alone 열만 선택한 후 값을 True로 변경하세요

In [7]:

mask4 = (titanic['age']<10)|(titanic['age']>=60), ['alone']
titanic4 = titanic.loc[mask4]  

print(titanic4.head(), '\n')
titanic4['alone'] = True

print(titanic4.head())



    alone
7   False
10  False
16  False
24  False
33   True 

    alone
7    True
10   True
16   True
24   True
33   True


# 과제 7/1

### Q11. 최대값 - 최소값을 반환하는 사용자 함수를 min_max로 정의하고 age, fare 열로 구성되는 타이타닉 데이터프레임df 에 적용하여 결과와 자료형을 출력하세요. source : titanic = sns.load_dataset('titanic')


In [15]:
df_af = df[['age','fare']]
min_max = lambda x:x.max()-x.min()
df_af = df_af.apply(min_max)
print(df_af, type(df_af))

age      79.5800
fare    512.3292
dtype: float64 <class 'pandas.core.series.Series'>


### Q12. df 에서 'age' 칼럼과 'fare'칼럼을 더한 가격을 리턴하는 사용자 함수를 정의한 후 df['add'] 변수를 생성하여 출력하세요.(lambda 함수 사용)


In [ ]:
#df['add'] = pd.concat([df['age'] + df['fare']], axis=1)

In [19]:
# def add(x):
#     return  a+b
# df['add'] = df..apply(f)
# df
# df['add'] = (df['age'] + df['fare'])
# df

# def add(x):
#     return x.age + x.fare
# df['add'] = add(df)
# df.head()

add = lambda x:x.age+x.fare
df1['add'] = add(df1)
df1

age     fare       add
0    22.0   7.2500   29.2500
1    38.0  71.2833  109.2833
2    26.0   7.9250   33.9250
3    35.0  53.1000   88.1000
4    35.0   8.0500   43.0500
..    ...      ...       ...
886  27.0  13.0000   40.0000
887  19.0  30.0000   49.0000
888   NaN  23.4500       NaN
889  26.0  30.0000   56.0000
890  32.0   7.7500   39.7500

[891 rows x 3 columns]

### Q13. 'age', 'fare' 두개 칼럼으로 구성되는 titanic 데이터프레임에서 [0,1]의 값을 NaN을 변경한 후 df 이름으로 출력하세요.


In [20]:
import numpy as np

In [21]:
titanic = df[['age','fare']]
titanic.head()

age     fare
0  22.0   7.2500
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500

In [24]:
import warnings
warnings.filterwarnings('ignore')
titanic.iloc[0,1] = np.nan
df = titanic
df.head()

age     fare
0  22.0      NaN
1  38.0  71.2833
2  26.0   7.9250
3  35.0  53.1000
4  35.0   8.0500


### Q14. 각 열의 NaN 찾기, 각 열의 NaN 개수 반환, 데이터프레임의 총 NaN 개수를 반환하는 사용자 함수를 정의한 후 df.pipe 메소드로 함수 매핑한 결과를 모두 출력하세요.


In [33]:
def a(x):
    x = titanic.isnull()
    y = titanic.isnull().sum()
    z = titanic.isnull().sum().sum()
    return print(x,'\n\n',y,'\n\n',z)

print(titanic.pipe(a))


       age   fare
0    False   True
1    False  False
2    False  False
3    False  False
4    False  False
..     ...    ...
886  False  False
887  False  False
888   True  False
889  False  False
890  False  False

[891 rows x 2 columns] 

 age     177
fare      1
dtype: int64 

 178
None


### Q15. 타이타닉 df의 열 순서를 아래와 같이 재구성하고 출력하세요
* 열 선택 : ['pclass', 'sex', 'age', 'survived'].
* 열 이름의 리스트 만들기
* 열 이름을 알파벳 순으로 정렬하기
* 열 이름을 기존 순서의 정반대 역순으로 정렬하기
* 열 이름을 사용자가 정의한 임의의 순서로 재배치하기


In [121]:
df = df[['pclass', 'sex', 'age', 'survived']]
# pclass_li = df.pclass.tolist() 이거 만들라는건가?
# #pclass_li
df_li = ['pclass', 'sex', 'age', 'survived']

df_li.sort()
df = df[[i for i in df_li]]

df_li.sort(reverse=True)
df = df[[i for i in df_li]]

df = df[['pclass', 'survived', 'sex','age']]

df.head()

# columns = ['pclass', 'survived', 'sex','age']
# df_custoumed = df[columns]
# df.head()

pclass  survived     sex   age
0       3         0    male  22.0
1       1         1  female  38.0
2       3         1  female  26.0
3       1         1  female  35.0
4       3         0    male  35.0

### 그룹 연산
- 복잡한 데이터를 어떤 기준에 따라 여러 그룹으로 나누어서 관찰할 수 있으며 이런 방식으로 분할 처리하는 것을 그룹 연산이라 함.
- 그룹연산은 데이터를 집계, 반환, 필터하는 효율적이며 판다스 groupby() 메소드를 사용
- 그룹 객체 만들기(분할)
- 그룹 연산 메소드(적용-결합)

In [34]:
titanic = sns.load_dataset('titanic')
df = titanic.loc[:,['age','sex','class','fare','survived']]

print('승객 수:', len(df)) # 인덱스 수
print(df.head())

승객 수: 891
    age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
1  38.0  female  First  71.2833         1
2  26.0  female  Third   7.9250         1
3  35.0  female  First  53.1000         1
4  35.0    male  Third   8.0500         0


In [24]:
df.columns

Index(['age', 'sex', 'class', 'fare', 'survived'], dtype='object')

In [35]:
# class열을 group으로 
grouped = df.groupby(['class'])
print(grouped)
list(grouped)

[('First',
        age     sex  class     fare  survived
  1    38.0  female  First  71.2833         1
  3    35.0  female  First  53.1000         1
  6    54.0    male  First  51.8625         0
  11   58.0  female  First  26.5500         1
  23   28.0    male  First  35.5000         1
  ..    ...     ...    ...      ...       ...
  871  47.0  female  First  52.5542         1
  872  33.0    male  First   5.0000         0
  879  56.0  female  First  83.1583         1
  887  19.0  female  First  30.0000         1
  889  26.0    male  First  30.0000         1
  
  [216 rows x 5 columns]),
 ('Second',
        age     sex   class     fare  survived
  9    14.0  female  Second  30.0708         1
  15   55.0  female  Second  16.0000         1
  17    NaN    male  Second  13.0000         1
  20   35.0    male  Second  26.0000         0
  21   34.0    male  Second  13.0000         1
  ..    ...     ...     ...      ...       ...
  866  27.0  female  Second  13.8583         1
  874  28.0  female

In [38]:
# 그룹 객체를 iteration으로 출력
for key, group in grouped:
    print('*key:', key)
    print('*number:', len(group))
    print(group.head())

*key: First
*number: 216
     age     sex  class     fare  survived
1   38.0  female  First  71.2833         1
3   35.0  female  First  53.1000         1
6   54.0    male  First  51.8625         0
11  58.0  female  First  26.5500         1
23  28.0    male  First  35.5000         1
*key: Second
*number: 184
     age     sex   class     fare  survived
9   14.0  female  Second  30.0708         1
15  55.0  female  Second  16.0000         1
17   NaN    male  Second  13.0000         1
20  35.0    male  Second  26.0000         0
21  34.0    male  Second  13.0000         1
*key: Third
*number: 491
    age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
2  26.0  female  Third   7.9250         1
4  35.0    male  Third   8.0500         0
5   NaN    male  Third   8.4583         0
7   2.0    male  Third  21.0750         0


In [41]:
# 연산 메서드 적용
average = grouped.mean()
average

age       fare  survived
class                                 
First   38.233441  84.154687  0.629630
Second  29.877630  20.662183  0.472826
Third   25.140620  13.675550  0.242363

In [7]:
stat = grouped.max()
stat

age   sex      fare  survived
class                                 
First   80.0  male  512.3292         1
Second  70.0  male   73.5000         1
Third   74.0  male   69.5500         1

In [8]:
# Q, 3rd 그룹만을 선택해서 group3 이름으로 저장하고 통계요약표를 출력하세요.
group3 = grouped.get_group('Third')
group3.describe()

age        fare    survived
count  355.000000  491.000000  491.000000
mean    25.140620   13.675550    0.242363
std     12.495398   11.778142    0.428949
min      0.420000    0.000000    0.000000
25%     18.000000    7.750000    0.000000
50%     24.000000    8.050000    0.000000
75%     32.000000   15.500000    0.000000
max     74.000000   69.550000    1.000000

In [9]:
grouped_two = df.groupby(['class','sex'])
average_two = grouped_two.mean()
average_two

age        fare  survived
class  sex                                    
First  female  34.611765  106.125798  0.968085
       male    41.281386   67.226127  0.368852
Second female  28.722973   21.970121  0.921053
       male    30.740707   19.741782  0.157407
Third  female  21.750000   16.118810  0.500000
       male    26.507589   12.661633  0.135447

In [10]:
group3f = grouped_two.get_group(('Third','female'))
group3f.head()

age     sex  class     fare  survived
2   26.0  female  Third   7.9250         1
8   27.0  female  Third  11.1333         1
10   4.0  female  Third  16.7000         1
14  14.0  female  Third   7.8542         0
18  31.0  female  Third  18.0000         0

In [11]:
# 모든 열에 여러 함수를 매핑 
agg_all = grouped.agg(['min','max'])
agg_all.head()

age           sex       fare           survived    
         min   max     min   max  min       max      min max
class                                                       
First   0.92  80.0  female  male  0.0  512.3292        0   1
Second  0.67  70.0  female  male  0.0   73.5000        0   1
Third   0.42  74.0  female  male  0.0   69.5500        0   1

In [12]:
# 각 열마다 다른 함수를 적용하여 집계
agg_sep = grouped.agg({'fare':['min','max'],'age':'mean'})
agg_sep.head()

fare                  age
        min       max       mean
class                           
First   0.0  512.3292  38.233441
Second  0.0   73.5000  29.877630
Third   0.0   69.5500  25.140620

In [17]:
# Q. 데이터 개수가 200개 이상인 그룹만을 필터링하여 데이터 프레임으로 반환하여 grouped_filter 이름으로 저장 후 처음 5개 행을 출력하세요
grouped_filter = grouped.filter(lambda x:len(x)>=200 )  # grouped가 클래스로 그룹바이된거라  len(x) 는 각 클래스의 개수
grouped_filter

age     sex  class     fare  survived
0    22.0    male  Third   7.2500         0
1    38.0  female  First  71.2833         1
2    26.0  female  Third   7.9250         1
3    35.0  female  First  53.1000         1
4    35.0    male  Third   8.0500         0
..    ...     ...    ...      ...       ...
885  39.0  female  Third  29.1250         0
887  19.0  female  First  30.0000         1
888   NaN  female  Third  23.4500         0
889  26.0    male  First  30.0000         1
890  32.0    male  Third   7.7500         0

[707 rows x 5 columns]

In [14]:
grouped_filter['class'].value_counts()  ## second는 200개가 안넘어서 제외

Third     491
First     216
Second      0
Name: class, dtype: int64

# 과제 7/1
Q. age열의 평균이 30보다 작은 그룹만을 필터링하여 데이터프레임으로 반환하여 age_filter 이름으로 저장 후 처음 5개 행을 출력하세요

In [43]:

age_filter = grouped.filter(lambda x: x['age'].mean() < 30)
age_filter.head()


age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
2  26.0  female  Third   7.9250         1
4  35.0    male  Third   8.0500         0
5   NaN    male  Third   8.4583         0
7   2.0    male  Third  21.0750         0

In [34]:
 age_filter['class'].value_counts() 

Third     491
Second    184
First       0
Name: class, dtype: int64

## pivot
피벗 테이블을 구성하는 4가지 요소(행인덱스, 열인덱스, 데이터 값, 데이터 집계함수)에 적용할 데이터프레임의 열을 각각 지정하여 함수의 인자로 전달

In [44]:
import seaborn as sns
titanic = sns.load_dataset('titanic')
df = titanic.loc[:,['age','sex','class','fare','survived']]
df.head()

age     sex  class     fare  survived
0  22.0    male  Third   7.2500         0
1  38.0  female  First  71.2833         1
2  26.0  female  Third   7.9250         1
3  35.0  female  First  53.1000         1
4  35.0    male  Third   8.0500         0

In [59]:
pdf1 = pd.pivot_table(df, index='class', columns='sex', values='age', aggfunc='mean') # aggfunc 나타내는 방법:평균
pdf1.head()

sex        female       male
class                       
First   34.611765  41.281386
Second  28.722973  30.740707
Third   21.750000  26.507589

In [61]:
# 행, 열, 값에 사용할 열을 2개 이상 지정 - index, values, aggfunc에 각각 2개 이상 지정
pdf2 = pd.pivot_table(df, index=['class','sex'], columns='survived', values=['age','fare'],aggfunc=['mean','max'])
pdf2

mean                                      max        \
                     age                   fare               age         
survived               0          1           0           1     0     1   
class  sex                                                                
First  female  25.666667  34.939024  110.604167  105.978159  50.0  63.0   
       male    44.581967  36.248000   62.894910   74.637320  71.0  80.0   
Second female  36.000000  28.080882   18.250000   22.288989  57.0  55.0   
       male    33.369048  16.022000   19.488965   21.095100  70.0  62.0   
Third  female  23.818182  19.329787   19.773093   12.464526  48.0  63.0   
       male    27.255814  22.274211   12.204469   15.579696  74.0  45.0   

                                 
                 fare            
survived            0         1  
class  sex                       
First  female  151.55  512.3292  
       male    263.00  512.3292  
Second female   26.00   65.0000  
       male     73.50   39.0000  
Third  female   69.55   31.3875  
       male     69.55   56.4958

In [2]:
# 결측값
df_t = sns.load_dataset('titanic')
df_t.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [3]:
df_t.embarked.value_counts()

S    644
C    168
Q     77
Name: embarked, dtype: int64

In [4]:
# 누락데이터의 앞의 값으로 대체
df_t.embarked.fillna(method='ffill', inplace = True)
df_t.embarked.value_counts()

S    644
C    169
Q     78
Name: embarked, dtype: int64

# 과제 7/1
- age 컬럼의 결측치를 적절한 값으로 대체하시고 그 근거를 설명하세요.
- deck 컬럼의 삭제 여부를 판단하고 근거를 설명하시오


In [65]:
import numpy as np
# 결측값 평균으로 대체할까?
df_t = sns.load_dataset('titanic')
df_t.isnull().sum()
df_t.age.replace(np.nan, df_t.age.mean(), inplace=True)
df_t.age 
# deck 삭제하는게 맞지 않나? 889중에 688이 널값인디?

survived         0
pclass           0
sex              0
age              0
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [ ]:
df.dropna(subset=['column'], inplace=True) #찾고
df.dropna(inplace=True)  # 컬럼만 x 전체
df.fillna(method='바꿀값', inplace=True)
df.column.replace(np.nan, df.column.mean(), inplace=True) #대체하고


In [ ]:
df_t.age.fillna(method='', inplace=True)
df_t.age.value_counts()

<img src = 'dataset/apply_applymap_pipe.jpg' STYLE='width:800px;'>

[pipe vs applymap vs apply]  

pipe와 applymap과의 차이는 반환하는 결과가 pipe가 더 다양하다는 것으로 applymap은 데이터프레임 원소에 적용하므로 데이터프레임으로만 리턴되지만 pipe는 사용자 함수의 리턴값에 따라 개별 값, 시리즈, 데이터프레임을 반환할 수 있다.  

현재 가장 많이 쓰이는 메서드는 apply인데 그 이유는 다양한 형태로 값을 입력 받을 수 있고, 반환하는 형태도 폭이 넓기 때문. 결과는 같아보이지만, 시리즈를 이용해 데이터프레임의 행과 열도 이용할 수 있으며 시리즈 그 자체도 함수로 활용할 수 있다. 

만약 데이터프레임에 함수를 적용해야한다면 리턴하는 형태가 많은 pipe 사용도 고려


In [35]:
import pandas as pd
df = pd.DataFrame({'Col1':[30, 40, 50, 60],'Col2':[23, 35, 65, 45], 'Col3':[85, 87, 90, 89]}, index=list('ABCD'))
df

Col1  Col2  Col3
A    30    23    85
B    40    35    87
C    50    65    90
D    60    45    89

In [78]:
df.apply(lambda x:x.sum(), axis=0) # 행 남아

Col1    180
Col2    168
Col3    351
dtype: int64

In [79]:
df.Col1.apply(lambda x:x+1)

A    31
B    41
C    51
D    61
Name: Col1, dtype: int64

In [80]:
df.Col1.map(lambda x:x+1)

A    31
B    41
C    51
D    61
Name: Col1, dtype: int64

In [83]:
df.applymap(lambda x:'%.2f' %x)  #데이터프레임에만 
# 포멧으로 소숫점 둘쩃자리까지 나타냄

Col1   Col2   Col3
A  30.00  23.00  85.00
B  40.00  35.00  87.00
C  50.00  65.00  90.00
D  60.00  45.00  89.00

In [36]:
# The pipe() method allows you to apply one or more functions to the DataFrame object.
#하나이상의 데이터프레임
import pandas as pd

def change_age(x):
  x["age"]=[10, 20, 30]
  return x

data = {
  "name": ["Sally", "Mary", "John"],
  "age": [50, 40, 30]
}

df = pd.DataFrame(data)
df

name  age
0  Sally   50
1   Mary   40
2   John   30

In [85]:
df.pipe(change_age)

name  age
0  Sally   10
1   Mary   20
2   John   30

In [47]:
# 중복
df = pd.DataFrame({'c1':['a','a','b','a','b'],'c2':[1,1,1,2,2],'c3':[1,1,2,2,2]})
df

c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

In [50]:
# 전체 행 데이터중에서 중복     선행행과 행 전체가 겹치는 것 찾아냄 
df_dup = df.duplicated()
df_dup

0    False
1     True
2    False
3    False
4    False
dtype: bool

In [44]:
# 특정 열(c2) 데이터중에서 중복
col_dup = df.c2.duplicated()
col_dup

0    False
1     True
2     True
3    False
4     True
Name: c2, dtype: bool

In [45]:
# drop_duplicates() : 중복 행 제거
df2 = df.drop_duplicates()
df2

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

In [46]:
df3 = df.drop_duplicates(subset=['c2','c3'])
df3

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2

In [45]:
# 데이터 단위 변경
auto_df = pd.read_csv('dataset/auto-mpg.csv')
auto_df.index

RangeIndex(start=0, stop=398, step=1)

# 과제 7/1
'?'가 있는 인덱스(불린인덱싱)를 찾아서 삭제하세요(행삭제)



In [47]:
sr_bool = auto_df['horsepower'] =='?'
sr_bool.sum()

6

In [423]:
q_index = auto_df[sr_bool].index
q_index

Int64Index([32, 126, 330, 336, 354, 374], dtype='int64')

In [424]:
auto_df.iloc[q_index]

mpg  cylinders  displacement horsepower  weight  acceleration  \
32   25.0          4          98.0          ?    2046          19.0   
126  21.0          6         200.0          ?    2875          17.0   
330  40.9          4          85.0          ?    1835          17.3   
336  23.6          4         140.0          ?    2905          14.3   
354  34.5          4         100.0          ?    2320          15.8   
374  23.0          4         151.0          ?    3035          20.5   

     model year  origin              car name  
32           71       1            ford pinto  
126          74       1         ford maverick  
330          80       2  renault lecar deluxe  
336          80       1    ford mustang cobra  
354          81       2           renault 18i  
374          82       1        amc concord dl

In [425]:
auto_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    398 non-null    object 
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model year    398 non-null    int64  
 7   origin        398 non-null    int64  
 8   car name      398 non-null    object 
dtypes: float64(3), int64(4), object(2)
memory usage: 28.1+ KB


In [19]:
auto_df.horsepower.replace('?', np.nan, inplace=True)
auto_df.isnull().sum()

mpg             0
cylinders       0
displacement    0
horsepower      6
weight          0
acceleration    0
model year      0
origin          0
car name        0
dtype: int64

In [17]:
# f = auto_df[auto_df=='?'].index
# auto_df.drop(f, inplace=True)
auto_df.dropna(subset=['horsepower'], inplace=True)

auto_df.index


Int64Index([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,
            ...
            388, 389, 390, 391, 392, 393, 394, 395, 396, 397],
           dtype='int64', length=392)

In [ ]:
df.dropna(subset=['column'], inplace=True) #찾고
df.dropna(inplace=True)  # 컬럼만 x 전체
df.fillna(method='바꿀값', inplace=True)
df.column.replace(np.nan, df.column.mean(), inplace=True) #대체하고


In [93]:
# mpg흫 kpl로 환산
mpg_to_kpl = 1.60934 / 3.78541
mpg_to_kpl

0.42514285110463595

In [102]:
# auto_kpl = round(auto_df['mpg']*mpg_to_kpl, 2)
auto_df['kpl'] = auto_df.mpg.apply(lambda x:round(x*mpg_to_kpl, 2))
auto_df[['mpg','kpl']]


mpg    kpl
0    18.0   7.65
1    15.0   6.38
2    18.0   7.65
3    16.0   6.80
4    17.0   7.23
..    ...    ...
393  27.0  11.48
394  44.0  18.71
395  32.0  13.60
396  28.0  11.90
397  31.0  13.18

[392 rows x 2 columns]

In [94]:
auto_df.horsepower.unique()

array(['130', '165', '150', '140', '198', '220', '215', '225', '190',
       '170', '160', '95', '97', '85', '88', '46', '87', '90', '113',
       '200', '210', '193', '?', '100', '105', '175', '153', '180', '110',
       '72', '86', '70', '76', '65', '69', '60', '80', '54', '208', '155',
       '112', '92', '145', '137', '158', '167', '94', '107', '230', '49',
       '75', '91', '122', '67', '83', '78', '52', '61', '93', '148',
       '129', '96', '71', '98', '115', '53', '81', '79', '120', '152',
       '102', '108', '68', '58', '149', '89', '63', '48', '66', '139',
       '103', '125', '133', '138', '135', '142', '77', '62', '132', '84',
       '64', '74', '116', '82'], dtype=object)

In [95]:
import numpy as np
auto_df.horsepower.replace('?', np.nan, inplace=True)
auto_df.dropna(subset=['horsepower'],axis=0, inplace=True)
auto_df.horsepower.isnull().sum()

0

In [99]:
auto_df.horsepower = auto_df.horsepower.astype('float')
auto_df.horsepower.dtypes

dtype('float64')

In [46]:
auto_df = pd.read_csv('dataset/auto-mpg.csv')


In [47]:
auto_df.origin.unique() #고유값

array([1, 3, 2], dtype=int64)

Q. 아래 사항을 처리 하세요.
* origin 열의 고유값을 출력하세요.
* 정수형 데이터를 문자형 데이터로 변환한 후 고유값을 출력하세요.(1:'USA', 2:'EU', 3:'JAPAN')


In [48]:
auto_df.replace({1:'USA', 2:'EU', 3:'JAPAN'}, inplace=True)


auto_df.origin.unique()

array(['USA', 'JAPAN', 'EU'], dtype=object)

In [49]:
auto_df.origin = auto_df.origin.astype('category')
auto_df.origin.dtypes

CategoricalDtype(categories=['EU', 'JAPAN', 'USA'], ordered=False)

In [50]:
auto_df.origin = auto_df.origin.astype('str')
auto_df.origin.dtypes

dtype('O')

In [68]:
auto_df = pd.read_pickle('dataset/auto-mpg.pkl')
auto_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    int64  
 5   acceleration  392 non-null    float64
 6   model year    392 non-null    int64  
 7   origin        392 non-null    int64  
 8   car name      392 non-null    object 
dtypes: float64(4), int64(4), object(1)
memory usage: 30.6+ KB


In [69]:
# horsepower를 3개의 bin으로 나누고 bin_nmaes = ['저출력', '보통출력','고출력']으로 출력하세요
import numpy as np
count, bin_dividers = np.histogram(auto_df.horsepower, bins=3) # bins 범주화 알아서 값으로 3등분 
print(bin_dividers)
bin_names = ['저출력', '보통출력','고출력']
auto_df['hp_bin'] = pd.cut(x=auto_df.horsepower, # 데이터 배열
                           bins=bin_dividers,  # 경계값 리스트
                           labels = bin_names, # bin 이름
                           include_lowest=True)  # 첫 경계값 포함
auto_df[['horsepower','hp_bin']].head()

[ 46.         107.33333333 168.66666667 230.        ]


horsepower hp_bin
0       130.0   보통출력
1       165.0   보통출력
2       150.0   보통출력
3       150.0   보통출력
4       140.0   보통출력

In [70]:
auto_df.horsepower.unique()

array([130., 165., 150., 140., 198., 220., 215., 225., 190., 170., 160.,
        95.,  97.,  85.,  88.,  46.,  87.,  90., 113., 200., 210., 193.,
       100., 105., 175., 153., 180., 110.,  72.,  86.,  70.,  76.,  65.,
        69.,  60.,  80.,  54., 208., 155., 112.,  92., 145., 137., 158.,
       167.,  94., 107., 230.,  49.,  75.,  91., 122.,  67.,  83.,  78.,
        52.,  61.,  93., 148., 129.,  96.,  71.,  98., 115.,  53.,  81.,
        79., 120., 152., 102., 108.,  68.,  58., 149.,  89.,  63.,  48.,
        66., 139., 103., 125., 133., 138., 135., 142.,  77.,  62., 132.,
        84.,  64.,  74., 116.,  82.])

In [73]:
# Q. horsepower 열의 최대값(.max())의 절대값(abs)으로 모든 데이터를 나누는 정규화를 수행한 후 통계요약표를 출력하세요.
auto_df.horsepower = auto_df.horsepower/abs(auto_df.horsepower.max())
auto_df[['horsepower']].describe()

horsepower
count  392.000000
mean     0.454215
std      0.167353
min      0.200000
25%      0.326087
50%      0.406522
75%      0.547826
max      1.000000

In [50]:
fifa = pd.read_csv('dataset/FIFA.csv', encoding='cp949')
fifa.head()

ID               Name  Age Nationality  Overall                 Club  \
0  158023           L. Messi   31   Argentina       94         FC Barcelona   
1   20801  Cristiano Ronaldo   33    Portugal       94             Juventus   
2  190871          Neymar Jr   26      Brazil       92  Paris Saint-Germain   
3  193080             De Gea   27       Spain       91    Manchester United   
4  192985       K. De Bruyne   27     Belgium       91      Manchester City   

  Preferred_Foot       Work_Rate Position  Jersey_Number  \
0           Left  Medium/ Medium       RF             10   
1          Right       High/ Low       ST              7   
2          Right    High/ Medium       LW             10   
3          Right  Medium/ Medium       GK              1   
4          Right      High/ High      RCM              7   

   Contract_Valid_Until Height  Weight_lb  Release_Clause   Value  Wage  
0                  2021    5'7        159          226500  110500   565  
1                  2022    6'2        183          127100   77000   405  
2                  2022    5'9        150          228100  118500   290  
3                  2020    6'4        168          138600   72000   260  
4                  2023   5'11        154          196400  102000   355

# 과제 7/4

In [51]:
# Q. height의 값을 cm로 변환하여 새로운 컬럼 'height_cm'를 생성하세요.
# def f(x):
#     y = x.replace("'", "")
#     return y
def f(x):
    return int(x[0])*30.48 + int(x[2:])*2.54
newfifa = fifa.Height.apply(f)
#int(fifa.Height[i][0])*30.48 + int(fifa.Height[i][2:])*2.54

#newfifa = newfifa.astype('int')


fifa['Height_cm'] = round(newfifa, 2)

fifa.head()

ID               Name  Age Nationality  Overall                 Club  \
0  158023           L. Messi   31   Argentina       94         FC Barcelona   
1   20801  Cristiano Ronaldo   33    Portugal       94             Juventus   
2  190871          Neymar Jr   26      Brazil       92  Paris Saint-Germain   
3  193080             De Gea   27       Spain       91    Manchester United   
4  192985       K. De Bruyne   27     Belgium       91      Manchester City   

  Preferred_Foot       Work_Rate Position  Jersey_Number  \
0           Left  Medium/ Medium       RF             10   
1          Right       High/ Low       ST              7   
2          Right    High/ Medium       LW             10   
3          Right  Medium/ Medium       GK              1   
4          Right      High/ High      RCM              7   

   Contract_Valid_Until Height  Weight_lb  Release_Clause   Value  Wage  \
0                  2021    5'7        159          226500  110500   565   
1                  2022    6'2        183          127100   77000   405   
2                  2022    5'9        150          228100  118500   290   
3                  2020    6'4        168          138600   72000   260   
4                  2023   5'11        154          196400  102000   355   

   Height_cm  
0     170.18  
1     187.96  
2     175.26  
3     193.04  
4     180.34

### 데이터프레임 결합
* 판다스 join() 메소드는 merge() 함수를 기반으로 만들어졌기 때문에 기본 작동 방식이 서로 비슷
* join() 메소드는 두 데이터프레임의 행 인덱스를 기준으로 결합하는 점에서 merge와 차이가 있으나 on=keys 옵션 설정하면 열을 기준으로 결합
##행인덱스 기준 join 


In [75]:
data1 = np.arange(1,21).reshape(4,5)
data2 = np.arange(11,31).reshape(5,4)
df1 = pd.DataFrame(data1, columns=list('abcde'))
df2 = pd.DataFrame(data2, columns=list('dehi'))
print(df1, '\n')
print(df2)

    a   b   c   d   e
0   1   2   3   4   5
1   6   7   8   9  10
2  11  12  13  14  15
3  16  17  18  19  20 

    d   e   h   i
0  11  12  13  14
1  15  16  17  18
2  19  20  21  22
3  23  24  25  26
4  27  28  29  30


In [76]:
df1.join(df2, lsuffix='_a', rsuffix='_b')

a   b   c  d_a  e_a  d_b  e_b   h   i
0   1   2   3    4    5   11   12  13  14
1   6   7   8    9   10   15   16  17  18
2  11  12  13   14   15   19   20  21  22
3  16  17  18   19   20   23   24  25  26

In [77]:
df1.join(df2, lsuffix='_a', rsuffix='_b', how='inner') # 디폴트 이너 작은 프레임에 맞춰짐

a   b   c  d_a  e_a  d_b  e_b   h   i
0   1   2   3    4    5   11   12  13  14
1   6   7   8    9   10   15   16  17  18
2  11  12  13   14   15   19   20  21  22
3  16  17  18   19   20   23   24  25  26

In [78]:
df1.join(df2, lsuffix='_a', rsuffix='_b', how='outer')  # 큰 프레임에 맞춰짐 ㅇㅇ

a     b     c   d_a   e_a  d_b  e_b   h   i
0   1.0   2.0   3.0   4.0   5.0   11   12  13  14
1   6.0   7.0   8.0   9.0  10.0   15   16  17  18
2  11.0  12.0  13.0  14.0  15.0   19   20  21  22
3  16.0  17.0  18.0  19.0  20.0   23   24  25  26
4   NaN   NaN   NaN   NaN   NaN   27   28  29  30

In [79]:
df1['ind'] = [1,2,3,4]
df2['ind'] = [1,2,3,4,5]
df1

a   b   c   d   e  ind
0   1   2   3   4   5    1
1   6   7   8   9  10    2
2  11  12  13  14  15    3
3  16  17  18  19  20    4

In [80]:
df1.set_index('ind').join(df2.set_index('ind'), lsuffix='_a', rsuffix='_b')


a   b   c  d_a  e_a  d_b  e_b   h   i
ind                                        
1     1   2   3    4    5   11   12  13  14
2     6   7   8    9   10   15   16  17  18
3    11  12  13   14   15   19   20  21  22
4    16  17  18   19   20   23   24  25  26

In [427]:
df_s1 = pd.read_excel('dataset/stock price.xlsx', engine='openpyxl', index_col='id')
df_s2 = pd.read_excel('dataset/stock valuation.xlsx', engine='openpyxl', index_col='id')
print(df_s1, '\n')
print(df_s2)

       stock_name          value   price
id                                      
128940       한미약품   59385.666667  421000
130960     CJ E&M   58540.666667   98900
138250      엔에스쇼핑   14558.666667   13200
139480        이마트  239230.833333  254500
142280     녹십자엠에스     468.833333   10200
145990        삼양사   82750.000000   82000
185750        종근당   40293.666667  100500
192400      쿠쿠홀딩스  179204.666667  177500
199800         툴젠   -2514.333333  115400
204210     모두투어리츠    3093.333333    3475 

             name           eps     bps        per       pbr
id                                                          
130960     CJ E&M   6301.333333   54068  15.695091  1.829178
136480         하림    274.166667    3551  11.489362  0.887074
138040    메리츠금융지주   2122.333333   14894   6.313806  0.899691
139480        이마트  18268.166667  295780  13.931338  0.860437
145990        삼양사   5741.000000  108090  14.283226  0.758627
161390      한국타이어   5648.500000   51341   7.453306  0.820007
181710  NHN엔터테인먼트 

In [83]:
df_j1 = df_s1.join(df_s2, how='left')
df_j1

stock_name          value   price    name           eps       bps  \
id                                                                         
128940       한미약품   59385.666667  421000     NaN           NaN       NaN   
130960     CJ E&M   58540.666667   98900  CJ E&M   6301.333333   54068.0   
138250      엔에스쇼핑   14558.666667   13200     NaN           NaN       NaN   
139480        이마트  239230.833333  254500     이마트  18268.166667  295780.0   
142280     녹십자엠에스     468.833333   10200     NaN           NaN       NaN   
145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090.0   
185750        종근당   40293.666667  100500     종근당   3990.333333   40684.0   
192400      쿠쿠홀딩스  179204.666667  177500     NaN           NaN       NaN   
199800         툴젠   -2514.333333  115400     NaN           NaN       NaN   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335.0   

              per       pbr  
id                           
128940        NaN       NaN  
130960  15.695091  1.829178  
138250        NaN       NaN  
139480  13.931338  0.860437  
142280        NaN       NaN  
145990  14.283226  0.758627  
185750  25.185866  2.470259  
192400        NaN       NaN  
199800        NaN       NaN  
204210  40.802348  0.651359

In [84]:
df_j2 = df_s1.join(df_s2, how='inner')
df_j2

stock_name          value   price    name           eps     bps  \
id                                                                       
130960     CJ E&M   58540.666667   98900  CJ E&M   6301.333333   54068   
139480        이마트  239230.833333  254500     이마트  18268.166667  295780   
145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090   
185750        종근당   40293.666667  100500     종근당   3990.333333   40684   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335   

              per       pbr  
id                           
130960  15.695091  1.829178  
139480  13.931338  0.860437  
145990  14.283226  0.758627  
185750  25.185866  2.470259  
204210  40.802348  0.651359

# 과제 7/4
Q. df_j2에서 중복된 칼럼을 삭제한 후 블린 인덱싱을 이용하여 eps가 3000 보다 적거나 stock_name이 이마트인 데이터를 선택하여 
데이터프레임을 생성하고 df_j3 이름으로 저장 및 출력하세요.(단, '<' 와 '==' 를 반드시 사용해야 함)


In [437]:
df_j3 = df_j2.drop('name', axis=1)
df_j3
df_j3 = df_j3[(df_j3['eps'] < 3000)|(df_j3['stock_name'] == '이마트')]
df_j3

stock_name          value   price           eps     bps        per  \
id                                                                          
139480        이마트  239230.833333  254500  18268.166667  295780  13.931338   
204210     모두투어리츠    3093.333333    3475     85.166667    5335  40.802348   

             pbr  
id                
139480  0.860437  
204210  0.651359

### 데이터프레임 병합
* merge() 함수는 sql의 join 명령과 비슷한 방식으로 어떤 기준에 의해 두 데이터프레임을 병합한는 개념
* 기준이 되는 열이나 인덱스를 키라고 부름
* 키가 되는 열이나 인덱스는 반드시 양쪽 데이터프레임에 모두 존재함


In [85]:
data1 = np.arange(1,21).reshape(4,5)
data2 = np.arange(11,31).reshape(5,4)
df1 = pd.DataFrame(data1, columns=list('abcde'))
df2 = pd.DataFrame(data2, columns=list('dehi'))
print(df1, '\n')
print(df2)

    a   b   c   d   e
0   1   2   3   4   5
1   6   7   8   9  10
2  11  12  13  14  15
3  16  17  18  19  20 

    d   e   h   i
0  11  12  13  14
1  15  16  17  18
2  19  20  21  22
3  23  24  25  26
4  27  28  29  30


In [86]:
df1['ind'] = [1,2,3,4]
df2['ind'] = [1,2,3,4,5]
df1

a   b   c   d   e  ind
0   1   2   3   4   5    1
1   6   7   8   9  10    2
2  11  12  13  14  15    3
3  16  17  18  19  20    4

In [87]:
pd.merge(df1,df2,on='ind', how='outer')

a     b     c   d_x   e_x  ind  d_y  e_y   h   i
0   1.0   2.0   3.0   4.0   5.0    1   11   12  13  14
1   6.0   7.0   8.0   9.0  10.0    2   15   16  17  18
2  11.0  12.0  13.0  14.0  15.0    3   19   20  21  22
3  16.0  17.0  18.0  19.0  20.0    4   23   24  25  26
4   NaN   NaN   NaN   NaN   NaN    5   27   28  29  30

In [90]:
df1.count() # 각 열의 컬럼을 세줌

a      4
b      4
c      4
d      4
e      4
ind    4
dtype: int64

# 과제 7/4
df1, df2에 대하여 다음을 수행하세요
- on='None'으로교집합을 만들어 출력하세요
- 첫번째 df(stock_name), 두번째 df(name)을 기준으로 합집합(left)을 만들어 출력하세요. 😒
- 오른쪽 df를 기준으로 합친 후 출력하세요.
- df1에서 가격이 50000 미만은 데이터만 출력하세요.
- df2에서 per가 30 보다 크고 pbr이 0.8 보다 큰 데이터를 모두 출력하세요.


In [459]:
df_s1 = pd.read_csv('dataset/stock_price.csv', encoding='cp949')
df_s2 = pd.read_csv('dataset/stock_valuation.csv', encoding='cp949')
df_s1.head()
df_s2.head()

id     name           eps     bps        per       pbr
0  130960   CJ E&M   6301.333333   54068  15.695091  1.829178
1  136480       하림    274.166667    3551  11.489362  0.887074
2  138040  메리츠금융지주   2122.333333   14894   6.313806  0.899691
3  139480      이마트  18268.166670  295780  13.931338  0.860437
4  145990      삼양사   5741.000000  108090  14.283226  0.758627

In [462]:
#- on='None'으로교집합을 만들어 출력하세요  : 논 전체를 기준 공통된것만 나옴  on=key 면 열기준으로 결합함  논이니까 행기준으로 결합한거임
pd.merge(df_s1,df_s2,  how='inner')

id stock_name          value   price    name           eps     bps  \
0  130960     CJ E&M   58540.666670   98900  CJ E&M   6301.333333   54068   
1  139480        이마트  239230.833300  254500     이마트  18268.166670  295780   
2  145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090   
3  185750        종근당   40293.666670  100500     종근당   3990.333333   40684   
4  204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335   

         per       pbr  
0  15.695091  1.829178  
1  13.931338  0.860437  
2  14.283226  0.758627  
3  25.185866  2.470259  
4  40.802348  0.651359

In [477]:
#- 첫번째 df(stock_name), 두번째 df(name)을 기준으로 합집합(left)을 만들어 출력하세요.
df_s1.set_index('stock_name').join(df_s2.set_index('name'), lsuffix='_s1', rsuffix='_s2', how='left')
#pd.merge(fd_s1,df_s2 how='left', lseft_on='stock_name', right_on='name')


id_s1          value   price     id_s2           eps       bps  \
stock_name                                                                    
한미약품        128940   59385.666670  421000       NaN           NaN       NaN   
CJ E&M      130960   58540.666670   98900  130960.0   6301.333333   54068.0   
엔에스쇼핑       138250   14558.666670   13200       NaN           NaN       NaN   
이마트         139480  239230.833300  254500  139480.0  18268.166670  295780.0   
녹십자엠에스      142280     468.833333   10200       NaN           NaN       NaN   
삼양사         145990   82750.000000   82000  145990.0   5741.000000  108090.0   
종근당         185750   40293.666670  100500  185750.0   3990.333333   40684.0   
쿠쿠홀딩스       192400  179204.666700  177500       NaN           NaN       NaN   
툴젠          199800   -2514.333333  115400       NaN           NaN       NaN   
모두투어리츠      204210    3093.333333    3475  204210.0     85.166667    5335.0   

                  per       pbr  
stock_name                       
한미약품              NaN       NaN  
CJ E&M      15.695091  1.829178  
엔에스쇼핑             NaN       NaN  
이마트         13.931338  0.860437  
녹십자엠에스            NaN       NaN  
삼양사         14.283226  0.758627  
종근당         25.185866  2.470259  
쿠쿠홀딩스             NaN       NaN  
툴젠                NaN       NaN  
모두투어리츠      40.802348  0.651359

In [457]:
#- 오른쪽 df를 기준으로 합친 후 출력하세요.
pd.merge(df_s1,df_s2, how='right')

id stock_name          value     price       name           eps  \
0  130960     CJ E&M   58540.666670   98900.0     CJ E&M   6301.333333   
1  136480        NaN            NaN       NaN         하림    274.166667   
2  138040        NaN            NaN       NaN    메리츠금융지주   2122.333333   
3  139480        이마트  239230.833300  254500.0        이마트  18268.166670   
4  145990        삼양사   82750.000000   82000.0        삼양사   5741.000000   
5  161390        NaN            NaN       NaN      한국타이어   5648.500000   
6  181710        NaN            NaN       NaN  NHN엔터테인먼트   2110.166667   
7  185750        종근당   40293.666670  100500.0        종근당   3990.333333   
8  204210     모두투어리츠    3093.333333    3475.0     모두투어리츠     85.166667   
9  207940        NaN            NaN       NaN   삼성바이오로직스   4644.166667   

      bps        per       pbr  
0   54068  15.695091  1.829178  
1    3551  11.489362  0.887074  
2   14894   6.313806  0.899691  
3  295780  13.931338  0.860437  
4  108090  14.283226  0.758627  
5   51341   7.453306  0.820007  
6   78434  30.755864  0.827447  
7   40684  25.185866  2.470259  
8    5335  40.802348  0.651359  
9   60099  89.790059  6.938551

In [450]:
#- df1에서 가격이 50000 미만은 데이터만 출력하세요.
df_s3 = df_s1[df_s1['price']<50000]
df_s3

id stock_name         value  price
2  138250      엔에스쇼핑  14558.666670  13200
4  142280     녹십자엠에스    468.833333  10200
9  204210     모두투어리츠   3093.333333   3475

In [451]:
#- df2에서 per가 30 보다 크고 pbr이 0.8 보다 큰 데이터를 모두 출력하세요.
df_s4 = df_s2[(df_s2['per']>30)&(df_s2['pbr']>0.8)] 
df_s4

id       name          eps    bps        per       pbr
6  181710  NHN엔터테인먼트  2110.166667  78434  30.755864  0.827447
9  207940   삼성바이오로직스  4644.166667  60099  89.790059  6.938551

# 과제 7/4

In [7]:

import pandas as pd
import seaborn as sns
titanic = sns.load_dataset('titanic')
df = titanic
df.head()


survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

### Q16. 함수 매핑을 이용하여 age에 각각 10을 곱하고 5를 더한 후 다시 더한 결과 값이 출력되게 하세요.


In [9]:
# 함수정의 : (a*b) +(a+ c)
def mul(a,b,c):
    return (a*b) + (a+c)

#f = lambda x:10*x+5
df.age = df.age.apply(mul, b=10, c=5)


#df = titanic.loc[:, ['age','fare']]

print(df[['age','fare']].head())


      age     fare
0  2722.0   7.2500
1  4658.0  71.2833
2  3206.0   7.9250
3  4295.0  53.1000
4  4295.0   8.0500


### Q17. titanic.fare 데이터를 소수 2째자리 형태로 모두 출력하세요.
- 7.2500 -> 7.25


In [483]:
def f(x):
    y = round(x,2)
    return y
df.fare = df.fare.apply(f)
df.fare.head()

0     7.25
1    71.28
2     7.92
3    53.10
4     8.05
Name: fare, dtype: float64

### Q18. df에서 fare와 age를 선택하여 100을 곱한 후 출력하세요


In [486]:
f = lambda x:x*100
df[['fare', 'age']] = df[['fare', 'age']].apply(f)
df.head()

survived  pclass     sex      age  sibsp  parch    fare embarked  class  \
0         0       3    male  22500.0      1      0   725.0        S  Third   
1         1       1  female  38500.0      1      0  7128.0        C  First   
2         1       3  female  26500.0      0      0   792.0        S  Third   
3         1       1  female  35500.0      1      0  5310.0        S  First   
4         0       3    male  35500.0      0      0   805.0        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

### Q19. join() 메소드는 두 데이터프레임의 행 인덱스를 기준으로 결합한다. 2개의 주식데이터를 가져와서 join() 메소드로 아래와 같이 결합한 후 다음 사항을 수행하세요.
- df1과 df2의 교집합만 출력되도록 결합하여 df4에 저장하고 출력
- df4에서 중복된 칼럼을 삭제한 후 블린 인덱싱을 이용하여 eps가 3000 보다 적거나 stock_name이 이마트인 데이터를 선택하여 데이터프레임을 생성하고 df5 이름으로 저장 및 출력하세요.(단, '<' 와 '==' 를 반드시 사용해야 함)


In [49]:
import pandas as pd

# 주식 데이터를 가져와서 데이터프레임 만들기
df1 = pd.read_excel('./dataset/stock price.xlsx', index_col='id')
df2 = pd.read_excel('./dataset/stock valuation.xlsx', index_col='id')

# 데이터프레임 결합(join)
# df1의 행 인덱스를 기준으로 결합하는 how='left'옵션이 기본 적용
df4 = df1.join(df2, how='inner')
df4


stock_name          value   price    name           eps     bps  \
id                                                                       
130960     CJ E&M   58540.666667   98900  CJ E&M   6301.333333   54068   
139480        이마트  239230.833333  254500     이마트  18268.166667  295780   
145990        삼양사   82750.000000   82000     삼양사   5741.000000  108090   
185750        종근당   40293.666667  100500     종근당   3990.333333   40684   
204210     모두투어리츠    3093.333333    3475  모두투어리츠     85.166667    5335   

              per       pbr  
id                           
130960  15.695091  1.829178  
139480  13.931338  0.860437  
145990  14.283226  0.758627  
185750  25.185866  2.470259  
204210  40.802348  0.651359

In [50]:
df4 = df4.drop('name', axis=1)
df5 = df4[(df4['eps']<3000)|(df4['stock_name']=='이마트')]
df5

stock_name          value   price           eps     bps        per  \
id                                                                          
139480        이마트  239230.833333  254500  18268.166667  295780  13.931338   
204210     모두투어리츠    3093.333333    3475     85.166667    5335  40.802348   

             pbr  
id                
139480  0.860437  
204210  0.651359

### Q20. 아래 df에서 중복 행을 제거한 후 df2에 저장하고 출력하세요.


In [21]:
import pandas as pd
df = pd.DataFrame({'c1':['a', 'a', 'b', 'a', 'b'],
                  'c2':[1, 1, 1, 2, 2],
                  'c3':[1, 1, 2, 2, 2]})
df


c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

In [22]:
df = df[df['c2']>1]
df

c1  c2  c3
3  a   2   2
4  b   2   2

In [3]:
df.duplicate()  # 중복열 찾기

AttributeError: 'DataFrame' object has no attribute 'duplicate'

In [4]:
df2 = df.drop_duplicates()  # 제거#
df2

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

In [51]:
col1 = np.random.uniform(10,20,size=10)  #10~20 사이 실수 10개
col2 = np.random.uniform(10,20,size=10)
sr1 = pd.Series(col1, name='col1')
sr2 = pd.Series(col2, name='col2')
df = pd.concat([sr1, sr2], axis=1)
df = df.round({'col1':2, 'col2':1})
df.head()

col1  col2
0  10.80  18.8
1  19.05  15.4
2  14.23  15.3
3  10.04  13.4
4  17.36  16.7

In [52]:
# col2를 정수로 변환하여 출력
df1 = df.copy()
df1 = df1.col2.astype('int')
df1.head()


0    18
1    15
2    15
3    13
4    16
Name: col2, dtype: int32

In [53]:
df2 = df.copy()
df2.iloc[1:3,:] = np.nan
df2.head()

col1  col2
0  10.80  18.8
1    NaN   NaN
2    NaN   NaN
3  10.04  13.4
4  17.36  16.7

In [1]:
# null값을 0으로 채우기
#df2.fillna(0, inplace=True)   
df2.replace(np.nan, 0, inplace=True)
df2 = df2.astype({'col1':'int','col2':'int'})
df2.head()

NameError: name 'df2' is not defined

In [144]:
df2['col0'] = list('abcdefghij')
df2.set_index('col0', inplace=True)
df2.head()

col1  col2
col0            
a       14    17
b        0     0
c        0     0
d       18    15
e       11    15

In [146]:
df2.isna().sum()

col1    0
col2    0
dtype: int64

In [147]:
df2 = df2.reset_index()
df2.head()

col0  col1  col2
0    a    14    17
1    b     0     0
2    c     0     0
3    d    18    15
4    e    11    15

In [148]:
df3 = df2.reindex(index=[1,2], columns=['col1','col2'])
df3

col1  col2
1     0     0
2     0     0

In [151]:
df4 = df2.reindex(index=[3,4], columns=['col1','col2'])
df4

col1  col2
3    18    15
4    11    15

In [ ]:
# Scaling : MinMaxScaling

In [157]:
from sklearn.datasets import load_boston
import warnings
warnings.filterwarnings('ignore')
boston = load_boston()
print(boston.keys())
# 타겟 : 주택가격
print(boston.feature_names)
df_boston = pd.DataFrame(boston.data, columns=boston.feature_names)
df_boston['target'] = boston.target
df_boston.head()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename', 'data_module'])
['CRIM' 'ZN' 'INDUS' 'CHAS' 'NOX' 'RM' 'AGE' 'DIS' 'RAD' 'TAX' 'PTRATIO'
 'B' 'LSTAT']


CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  target  
0     15.3  396.90   4.98    24.0  
1     17.8  396.90   9.14    21.6  
2     17.8  392.83   4.03    34.7  
3     18.7  394.63   2.94    33.4  
4     18.7  396.90   5.33    36.2

In [162]:
from sklearn.preprocessing import MinMaxScaler  # 민맥스가 0,1

scaler = MinMaxScaler()
df_boston['RM'] = scaler.fit_transform(df_boston[['RM']])  #fit 준비단계   transform 
df_boston[['RM']].describe()

RM
count  506.000000
mean     0.521869
std      0.134627
min      0.000000
25%      0.445392
50%      0.507281
75%      0.586798
max      1.000000

In [164]:
from sklearn.preprocessing import StandardScaler   #평균 0 표준편차 
scaler = StandardScaler()
df_boston['INDUS'] = scaler.fit_transform(df_boston[['INDUS']])
df_boston[['INDUS']].describe()

INDUS
count  5.060000e+02
mean   2.804081e-16
std    1.000990e+00
min   -1.557842e+00
25%   -8.676906e-01
50%   -2.110985e-01
75%    1.015999e+00
max    2.422565e+00

In [10]:
# encoding: LavledEncoder, OndeHotEncoder

In [94]:
titanic = sns.load_dataset('titanic')
# 특정 컬럼의 값을 변경
titanic['sex'] = titanic.sex.replace({'male':0,'female':1})
# 열이름 변경
titanic.rename(columns={'sex':'gender','age':'age_cat','class':'c_level'}, inplace=True)
# 컬럼명을 대문자로 변경
titanic.columns = map(lambda x:str(x).upper(), titanic.columns)
# 컬럼명을 소문자로 변경
titanic.rename(columns=str.lower, inplace=True)
df = titanic[['survived','pclass','gender','c_level']]
df.head()


survived  pclass  gender c_level
0         0       3       0   Third
1         1       1       1   First
2         1       3       1   Third
3         1       1       1   First
4         0       3       0   Third

In [95]:
import warnings
warnings.filterwarnings('ignore')
df['gender'].replace([1,0],['female','male'], inplace=True)
df.head()

survived  pclass  gender c_level
0         0       3    male   Third
1         1       1  female   First
2         1       3  female   Third
3         1       1  female   First
4         0       3    male   Third

In [96]:
df1 = df.copy()

In [97]:
# 인코딩 - LabelEncoder    # 숫자가 커지면 회귀에 있어서 라벨인코더 문제생김 
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
features = ['gender','c_level']
for f in features:
    df1[f] = le.fit_transform(df[f])
df1.head()

survived  pclass  gender  c_level
0         0       3       1        2
1         1       1       0        0
2         1       3       0        2
3         1       1       0        0
4         0       3       1        2

In [ ]:
# 인코딩 - OneHotEncoder 희소행렬  데이터용량이 커짐  컬럼 값의 종류에 따라 
pd.get_dummise()

In [98]:
df2 = df.copy()
df3 = df2[['gender','c_level']]
pd.get_dummies(df3).head()

gender_female  gender_male  c_level_First  c_level_Second  c_level_Third
0              0            1              0               0              1
1              1            0              1               0              0
2              1            0              0               0              1
3              1            0              1               0              0
4              0            1              0               0              1

# 과제 7/5
- titanic 데이터셋에서 embarked, who, adult_male에 대하여 LabelEncoding 하세요
- titanic 데이터셋에서 embarked, who, adult_male에 대하여 OneHotEncoding 하세요

In [11]:
titanic = sns.load_dataset('titanic')
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
df = titanic.copy()
features = ['embarked', 'who', 'adult_male']
for f in features:
    df[f] = le.fit_transform(df[f])
df.head()

survived  pclass     sex   age  sibsp  parch     fare  embarked  class  \
0         0       3    male  22.0      1      0   7.2500         2  Third   
1         1       1  female  38.0      1      0  71.2833         0  First   
2         1       3  female  26.0      0      0   7.9250         2  Third   
3         1       1  female  35.0      1      0  53.1000         2  First   
4         0       3    male  35.0      0      0   8.0500         2  Third   

   who  adult_male deck  embark_town alive  alone  
0    1           1  NaN  Southampton    no  False  
1    2           0    C    Cherbourg   yes  False  
2    2           0  NaN  Southampton   yes   True  
3    2           0    C  Southampton   yes  False  
4    1           1  NaN  Southampton    no   True

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     891 non-null    int32   
 8   class        891 non-null    category
 9   who          891 non-null    int32   
 10  adult_male   891 non-null    int64   
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(1), category(2), float64(2), int32(2), int64(5), object(3)
memory usage: 79.8+ KB


In [13]:
titanic = sns.load_dataset('titanic')
titanic.adult_male = titanic.adult_male.astype('str')
df1 = titanic.copy()
df2 = df1[['embarked', 'who', 'adult_male']]
pd.get_dummies(df2).head()

embarked_C  embarked_Q  embarked_S  who_child  who_man  who_woman  \
0           0           0           1          0        1          0   
1           1           0           0          0        0          1   
2           0           0           1          0        0          1   
3           0           0           1          0        0          1   
4           0           0           1          0        1          0   

   adult_male_False  adult_male_True  
0                 0                1  
1                 1                0  
2                 1                0  
3                 1                0  
4                 0                1

In [63]:
# Q 20 데이터
import pandas as pd
df = pd.DataFrame({'c1':['a', 'a', 'b', 'a', 'b'],
                  'c2':[1, 1, 1, 2, 2],
                  'c3':[1, 1, 2, 2, 2]})
print(df)


  c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2


### Q21. Q20의 df에서 drop_duplicates() 메소드를 사용하여 c2, c3열을 기준으로 중복 행을 제거한 후 df3에 저장하고 출력하세요.


In [67]:
df3 = df[['c2','c3']].duplicated()
print(df3)
df3 = df.drop_duplicates(subset=['c2','c3'])
print(df3)



0    False
1     True
2    False
3    False
4     True
dtype: bool
  c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2


### Q33. './dataset/stock-data.csv'를 데이터프레임으로 불러와서 datetime64 자료형으로 변환한 후에 년, 월, 일로 분리하고 year를 인덱스로 셋팅하여 출력하세요.


In [26]:
from datetime import datetime
import time
import numpy as np
sd = pd.read_csv('dataset/stock-data.csv')
sd.head()
sd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Date    20 non-null     object
 1   Close   20 non-null     int64 
 2   Start   20 non-null     int64 
 3   High    20 non-null     int64 
 4   Low     20 non-null     int64 
 5   Volume  20 non-null     int64 
dtypes: int64(5), object(1)
memory usage: 1.1+ KB


In [160]:
#sd['Date'] = pd.to_datetime(sd['Date'])
sd['Date'] = sd.Date.astype('datetime64')
sd.Date.info()
sd['year'] = sd['Date'].dt.year
sd.drop('Date', axis=1, inplace=True)
sd.set_index('year', inplace=True)
sd.head()

<class 'pandas.core.series.Series'>
RangeIndex: 20 entries, 0 to 19
Series name: Date
Non-Null Count  Dtype         
--------------  -----         
20 non-null     datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 288.0 bytes


Close  Start   High    Low  Volume
year                                    
2018  10100  10850  10900  10000  137977
2018  10700  10550  10900   9990  170253
2018  10400  10900  10950  10150  155769
2018  10900  10800  11050  10500  133548
2018  10800  10900  11000  10700   63039

In [ ]:
sd['year'] = sd['Date'].dt.year
sd['month'] = sd['Date'].dt.month
sd['day'] = sd['Date'].dt.day

### Q34. './dataset/stock-data.csv'를 데이터프레임으로 불러와서 datetime64 자료형으로 변환한 후에 Period로 변환하고 년월일 표기 변경하여 년월을 인덱스로 셋팅하여 출력하세요.


In [27]:
sd = pd.read_csv('dataset/stock-data.csv')
sd.head()
# sd['Date'] = pd.to_datetime(sd['Date'])
sd['Date'] = sd.Date.astype('datetime64')
sd.Date.info()

<class 'pandas.core.series.Series'>
RangeIndex: 20 entries, 0 to 19
Series name: Date
Non-Null Count  Dtype         
--------------  -----         
20 non-null     datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 288.0 bytes


In [24]:
sd.set_index('Date', inplace=True)
sd.index = sd.index.to_period(freq='m')
sd.head()

Close  Start   High    Low  Volume
Date                                       
2018-07  10100  10850  10900  10000  137977
2018-06  10700  10550  10900   9990  170253
2018-06  10400  10900  10950  10150  155769
2018-06  10900  10800  11050  10500  133548
2018-06  10800  10900  11000  10700   63039

In [29]:
sd = pd.read_csv('dataset/stock-data.csv')
sd.head()

sd['Date'] = sd.Date.astype('datetime64')
sd.Date.info()

#df['New_Date'] = pd.to_datetime(df['Date'])
sd['year'] = sd['Date'].dt.to_period(freq = 'Y')
sd['month'] = sd['Date'].dt.to_period(freq = 'M')
sd['day'] = sd['Date'].dt.to_period(freq = 'D')
sd = sd.set_index(['year','month','day'])





sd.drop('Date', axis=1, inplace=True)
sd.head()

<class 'pandas.core.series.Series'>
RangeIndex: 20 entries, 0 to 19
Series name: Date
Non-Null Count  Dtype         
--------------  -----         
20 non-null     datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 288.0 bytes


Close  Start   High    Low  Volume
year month   day                                           
2018 2018-07 2018-07-02  10100  10850  10900  10000  137977
     2018-06 2018-06-29  10700  10550  10900   9990  170253
             2018-06-28  10400  10900  10950  10150  155769
             2018-06-27  10900  10800  11050  10500  133548
             2018-06-26  10800  10900  11000  10700   63039

In [1]:
import pandas as pd

In [2]:
dir(pd)

['BooleanDtype',
 'Categorical',
 'CategoricalDtype',
 'CategoricalIndex',
 'DataFrame',
 'DateOffset',
 'DatetimeIndex',
 'DatetimeTZDtype',
 'ExcelFile',
 'ExcelWriter',
 'Flags',
 'Float32Dtype',
 'Float64Dtype',
 'Float64Index',
 'Grouper',
 'HDFStore',
 'Index',
 'IndexSlice',
 'Int16Dtype',
 'Int32Dtype',
 'Int64Dtype',
 'Int64Index',
 'Int8Dtype',
 'Interval',
 'IntervalDtype',
 'IntervalIndex',
 'MultiIndex',
 'NA',
 'NaT',
 'NamedAgg',
 'Period',
 'PeriodDtype',
 'PeriodIndex',
 'RangeIndex',
 'Series',
 'SparseDtype',
 'StringDtype',
 'Timedelta',
 'TimedeltaIndex',
 'Timestamp',
 'UInt16Dtype',
 'UInt32Dtype',
 'UInt64Dtype',
 'UInt64Index',
 'UInt8Dtype',
 '__all__',
 '__builtins__',
 '__cached__',
 '__deprecated_num_index_names',
 '__dir__',
 '__doc__',
 '__docformat__',
 '__file__',
 '__getattr__',
 '__git_version__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_config',
 '_is_numpy_dev',
 '_libs',
 '_testing',
 '_typing',
 '_ve